### PR TITLE
[PROGRAMMES-6351] - Added UAS module to send SMP player activities

### PR DESCRIPTION
**Pull request by tatitati**
## UAS module Specs:
- [x] UAS service only send data to Activity server when the user is logged in (by default), but it includes some parameters to send data when the user is also logged out (configurable in code).
These requests look like this:

| Code response | Method | Type | Protocol | Url | Call |
|---|---|---|---|---|---|
| 202 | POST | plays | https | activity.test.api.bbc.co.uk | xhr |
| 202 | POST | heartbeat | https | heartbeat.activity.test.api.bbc.co.uk | xhr |
| 202 | POST | reads | https | activity.test.api.bbc.co.uk | xhr |

- [x] UAS send different activities: **started, paused, ended, reads, heartbeat**. Example:
```json
{
"action" : "started",
"actionContext" : "urn:bbc:tv:version_offset:p06f8f0f#22",
"resourceDomain" : "tv",
"resourceId" : "p06f8f0b",
"resourceType" : "clip",
} 
```
- [x] "Heartbeat" event is sent every 30 seconds
- [x] "Reads" event is sent when the user visit the page through a recommendation and is loggedin
- [x] UAS only runs in clip page.
- [x] UAS only send data for int/test/stage/live environments.
- [x] Uas stop sending data to Activity server for a few minutes if it receives error responses (A.K.A circuitbreaker in js)

## To test it:
1. Go to an INT v3 clip page like: https://programmes-frontend.int.api.bbc.co.uk/programmes/p06f8f0b?__flush_cache=1
2. Login
3. refresh the page after login. You should see a the next requests similar to the table above.

## Question:
Uas module use an ApiKey when sending activities to the server:
https://github.com/bbc/programmes-frontend/blob/uas-module/src/Ds2013/Presenters/Utilities/SMP/SmpPresenter.php#L132
Not sure if is useful to store it in any secure place as is gonna be dumped later in clear text in the javascript DOM. In V2 this is the case.

<details>
  <summary>Expand information about the implementation of UAS module:</summary>

I tried to find an existing UAS client that we could use. However I was no lucky. As a prove of this you can see some interesting options I found, but all of them were not reliable as they are done in scala, node, etc:
* https://github.com/bbc/rms-uas-client
* https://github.com/bbc/ibl-user-activity
* https://github.com/bbc/ibl-user-activity-client
* https://github.com/bbc/node-uas-client

From my conversations with some teams I got:
- IBL team doesn't use any UAS client, instead they send data to IBL, and IBL transform it into UAS format. However /programmes wouldn't be able to use IBL as it transform into an IPlayer specific way.
- Iplayer team use a backend UAS client only usable by Java or scala systems.

UAS client Confluence page says that we can just require it on the fly, this might be a reliable solution), however I didn't do it as I assume there is a reason to not be required in V2:
https://confluence.dev.bbc.co.uk/display/activity/User+Activity+Client+Library+User+Guide#UserActivityClientLibraryUserGuide-Gettingstarted

I finished choosing the existing V2:  https://github.com/bbc/uas-client-pal
</details>

<details>
  <summary>How I review this?</summary>
  In order to make a review you must know that the only file I could refactor (clean and improve properly) was UasService.js
This file acts as a façade to UAS module and as you can imagine this UasService relies in the UasClient to send activity user.
</details>

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: assets/js/smp/smp.js [hunk 1]
```diff
@@ -145,9 +145,28 @@ define(['jquery-1.9', 'smp/smpStatePersistance', 'bump-3'], function ($, SmpStat
         var listenProgressTimeEvents = function ()
         {
             var savedTime = 0;
+            var time = 0;
```
Comment (votemike) [inline assets/js/smp/smp.js]: What's time? The current time? The time the user started listening? The time the media was first broadcast?

File: assets/js/uas/corsRequest.js [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+                env = '';
+            }
+
+            if (optionsForRequest.userIsNotSignedIn) {
+                baseUrl = 'https://nsi.activity.';
+            } else {
+                baseUrl = 'https://activity.';
+            }
+
+            baseUrl = baseUrl + env + 'api.bbc.' + this.config.domainSuffix + '/my/' + optionsForRequest.activityType;
```
Comment (votemike) [inline assets/js/uas/corsRequest.js]: `baseUrl += env + 'api.bbc.' + this.config.domainSuffix + '/my/' + optionsForRequest.activityType;`?

File: assets/js/uas/corsRequest.js [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+            if (this.config.baseUri) {
+                baseUrl = this.config.baseUri + '/my/' + optionsForRequest.activityType;
```
Comment (votemike) [inline assets/js/uas/corsRequest.js]: Line 54 to 60 were just spent constructing the baseUrl, now you're overwriting it. Is there a way to simplify or refactor this logic?

File: assets/js/uas/uasService.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+                    uasData.metaData = JSON.stringify(self.recommendationData);
+                }
+
+                sendToUas(uasData);
+            }
+        };
+
+        // private
+
+        var self = this;
```
Comment (votemike) [inline assets/js/uas/uasService.js]: I like that sometimes we use `self` and sometimes we use `_this`

File: src/Controller/FindByPid/ClipController.php [hunk 5]
```diff
@@ -40,7 +41,8 @@ public function __invoke(
         RelatedLinksService $relatedLinksService,
         SegmentEventsService $segmentEventsService,
         StructuredDataHelper $structuredDataHelper,
-        VersionsService $versionsService
+        VersionsService $versionsService,
+        CosmosInfo $cosmosInfo
```
Comment (votemike) [inline src/Controller/FindByPid/ClipController.php]: All of the other params coming though seem alphabetised, shall we keep it that way?

File: src/Ds2013/Presenters/Utilities/SMP/SmpPresenter.php [hunk 6]
```diff
@@ -128,9 +134,10 @@ public function getFactoryOptions(): array
      */
     private function getUasConfig(): array
     {
+
         return [
-            'apiKey' => '', // @todo
-            'env' => '', // @todo
+            'apiKey' => 'bapd63mcqopnp', // @todo
```
Comment (votemike) [inline src/Ds2013/Presenters/Utilities/SMP/SmpPresenter.php]: This isn't a LIVE key is it? As now it's in a public repo.

Comment (Tasiobg) [reply src/Ds2013/Presenters/Utilities/SMP/SmpPresenter.php]: Francisco mention there was a key in the javascript PAL library so he decided make it public here as well, in other words, for v2 this key can be seen in the javascript.

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: assets/js/smp/smp.js [hunk 7]
```diff
@@ -145,33 +145,56 @@ define(['jquery-1.9', 'smp/smpStatePersistance', 'bump-3'], function ($, SmpStat
         var listenProgressTimeEvents = function ()
         {
             var savedTime = 0;
+            var currentInstantInReproduction = 0;
```
Comment (BjornTwachtmann) [inline assets/js/smp/smp.js]: currentTime?

File: src/Controller/FindByPid/ClipController.php [hunk 8]
```diff
@@ -119,6 +121,7 @@ public function __invoke(
             'downloadableVersion' => $linkedVersions['downloadableVersion'],
             'streamableVersion' => $linkedVersions['streamableVersion'],
             'segmentEvents' => $segmentEvents,
+            'cosmosInfo' => $cosmosInfo,
```
Comment (BjornTwachtmann) [inline src/Controller/FindByPid/ClipController.php]: Avoid passing this into the view. Pass it into the constructor for the presenter factory as it's known at compile time.

File: templates/base_ds2013.html.twig [hunk 9]
```diff
@@ -46,7 +46,12 @@
                     'bump-3': "//emp.bbci.co.uk/emp/bump-3/bump-3",
                     'smp/smpFactory': "{{ asset_js("js/smp/smpFactory.js") }}",
                     'smp/smp': "{{ asset_js("js/smp/smp.js") }}",
-                    'smp/recommendationsBump': "{{ asset_js("js/smp/recommendationsBump.js") }}"
+                    'smp/recommendationsBump': "{{ asset_js("js/smp/recommendationsBump.js") }}",
+                    'uas/uasService': "{{ asset_js("js/uas/uasService.js") }}",
+                    'uasclient': "https://static.bbc.co.uk/uasclient/0.1.42/modules/uasclient",
+                    'idcta/idcta-1': "https://mybbc.files.bbci.co.uk/s/id/account-idcta/1.8.4/modules/idcta-v2/idcta-1",
```
Comment (BjornTwachtmann) [inline templates/base_ds2013.html.twig]: You shouldn't need to add this.

File: templates/base_ds2013.html.twig [hunk 10]
```diff
@@ -46,7 +46,12 @@
                     'bump-3': "//emp.bbci.co.uk/emp/bump-3/bump-3",
                     'smp/smpFactory': "{{ asset_js("js/smp/smpFactory.js") }}",
                     'smp/smp': "{{ asset_js("js/smp/smp.js") }}",
-                    'smp/recommendationsBump': "{{ asset_js("js/smp/recommendationsBump.js") }}"
+                    'smp/recommendationsBump': "{{ asset_js("js/smp/recommendationsBump.js") }}",
+                    'uas/uasService': "{{ asset_js("js/uas/uasService.js") }}",
```
Comment (BjornTwachtmann) [inline templates/base_ds2013.html.twig]: change the path for this, make it live under SMP I think

File: assets/js/uas/uasService.js [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+                600000  // 10 minutes
+            ]
+        }, uasConfig);
+
+        this.init = function() {
+            if (this.options.apiKey && userIsLoggedInAndHasTrackingEnabled()) {
+                var self = this;
+                istats.isReady(function() {
+                    self.recommendationData = getRecommendationDataFromIstats();
+                    require (['uasclient'], function(UasClient) {
```
Comment (BjornTwachtmann) [inline assets/js/uas/uasService.js]: Maybe move the require for uasclient to the initial define() block at the top?

File: assets/js/smp/smp.js [hunk 12]
```diff
@@ -41,7 +41,7 @@ define(['jquery-1.9', 'smp/smpStatePersistance', 'bump-3'], function ($, SmpStat
                     volume: null,
                     autoplay: false
                 },
-                rememberResume: true,
+                rememberResume: false,
```
Comment (tatitati) [inline assets/js/smp/smp.js]: This was wrong. By default is false in V2:
https://github.com/bbc/programmes/blob/develop/webapp/static-versioned/js-modules/smp/init.js#L61

File: assets/js/smp/smpStatePersistance.js [hunk 13]
```diff
@@ -54,25 +54,25 @@ define(['jquery-1.9','smp/cookieStore'], function ($, DataStore) {
...
         this.updateTimeResume = function (timestamp) {
-            dataStore.persist(COOKIE_RESUME, replaceTimeResumeForProgramme(timestamp));
+            replaceTimeResumeForProgrammeInCache(timestamp);
+            dataStore.persist(COOKIE_RESUME, cache);
         };
 
         this.removeTimeResume = function () {
-            dataStore.persist(COOKIE_RESUME, removeProgrammeFromCache());
+            removeProgrammeFromCache();
+            dataStore.persist(COOKIE_RESUME, cache);
```
Comment (tatitati) [inline assets/js/smp/smpStatePersistance.js]: Fixed BUG. removeProgrammeFromCache() return null. When the reproduction was ended we were storing a null into the cookie

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

**Pull request merged**